pkg/controller/staticpod/staticpod_controller.go

```diff
@@ -0,0 +1,671 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an AS IS BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package staticpod
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kerr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appconfig "github.com/openyurtio/openyurt/cmd/yurt-manager/app/config"
+	appsv1alpha1 "github.com/openyurtio/openyurt/pkg/apis/apps/v1alpha1"
+	"github.com/openyurtio/openyurt/pkg/controller/staticpod/config"
+	"github.com/openyurtio/openyurt/pkg/controller/staticpod/upgradeinfo"
+	"github.com/openyurtio/openyurt/pkg/controller/staticpod/util"
+	utilclient "github.com/openyurtio/openyurt/pkg/util/client"
+	utildiscovery "github.com/openyurtio/openyurt/pkg/util/discovery"
+)
+
+func init() {
+	flag.IntVar(&concurrentReconciles, "staticpod-workers", concurrentReconciles, "Max concurrent workers for StaticPod controller.")
+}
+
+var (
+	concurrentReconciles      = 3
+	controllerKind            = appsv1alpha1.SchemeGroupVersion.WithKind("StaticPod")
+	True                      = true
+	upgradeSuccessCondition   = util.NewStaticPodCondition(appsv1alpha1.StaticPodUpgradeSuccess, corev1.ConditionTrue, "", "")
+	upgradeExecutingCondition = util.NewStaticPodCondition(appsv1alpha1.StaticPodUpgradeExecuting, corev1.ConditionTrue, "", "")
+)
+
+const (
+	controllerName = "StaticPod-controller"
+
+	StaticPodHashAnnotation     = "openyurt.io/static-pod-hash"
+	OTALatestManifestAnnotation = "openyurt.io/ota-latest-version"
+
+	hostPathVolumeName       = "hostpath"
+	hostPathVolumeMountPath  = "/etc/kubernetes/manifests/"
+	configMapVolumeName      = "configmap"
+	configMapVolumeMountPath = "/data"
+	hostPathVolumeSourcePath = hostPathVolumeMountPath
+
+	// UpgradeWorkerPodPrefix is the name prefix of worker pod which used for static pod upgrade
+	UpgradeWorkerPodPrefix     = "yurt-static-pod-upgrade-worker-"
+	UpgradeWorkerContainerName = "upgrade-worker"
+	UpgradeWorkerImage         = "openyurt/yurt-static-pod-upgrade:latest"
```

Comment:
how about add a new parameter of yurt-manager for setting UpgradeWorkerImage? and `openyurt/yurt-static-pod-upgrade:latest` can be a default value.

Reply:
done